packages/web/src/pages/sign-up-page/SignUpRootPage.tsx

```diff
@@ -1,60 +1,128 @@
+    allowedRoutes.push(SignUpPath.pickHandle)
+  }
+  if (signUpState.handle.value) {
+    // Already have handle
+    allowedRoutes.push(SignUpPath.finishProfile)
+  }
+  if (signUpState.name.value) {
+    // Already have display name
+    // At this point the account is fully created & logged in; now user can't back to account creation steps
+    allowedRoutes = [SignUpPath.selectGenres]
+  }
 
+  // TODO: These checks below here may need to fall under a different route umbrella separate from sign up
+  if (signUpState.genres) {
+    // Already have genres selected
+    allowedRoutes.push(SignUpPath.selectArtists)
+  }
+
+  if (signUpState.followArtists?.selectedUserIds?.length >= 3) {
+    // Already have 3 artists followed
+    // Done with sign up if at this point so we return early (none of these routes are allowed anymore)
+    // TODO: trigger welcome modal when redirecting from here
+    return { isAllowedRoute: false, correctedRoute: TRENDING_PAGE }
+  }
+
+  const isAllowedRoute = allowedRoutes.includes(attemptedPath)
+  // If requested route is allowed return that, otherwise return the last step in the route stack
+  const correctedPath = isAllowedRoute
+    ? attemptedPath
+    : allowedRoutes[allowedRoutes.length - 1]
+  return {
+    isAllowedRoute,
+    correctedRoute: `/signup/${correctedPath}`
+  }
+}
+
+/**
+ * <Route> wrapper that handles redirecting through the sign up page flow
+ */
+export function SignUpRoute({ children, ...rest }: RouteProps) {
+  const existingSignUpState = useSelector((state: AppState) => getSignOn(state))
+  return (
+    <Route
+      {...rest}
+      render={({ location }) => {
+        // Check if the route is allowed, if not we redirect accordingly
+        const { isAllowedRoute, correctedRoute } = determineAllowedRoute(
+          existingSignUpState,
+          location.pathname
+        )
+        return isAllowedRoute ? (
+          <>{children}</>
+        ) : (
+          <Redirect to={correctedRoute} />
+        )
+      }}
+    />
+  )
+}
+
+export const SignUpRootPage = () => {
   return (
     <div>
-      {signUpState.stage === 'sign-up' ? (
-        <SignUpPage onNext={setSignUpState} />
-      ) : null}
-      {signUpState.stage === 'create-password' ? (
-        <CreatePasswordPage
-          onPrevious={setSignUpState}
-          onNext={setSignUpState}
-        />
-      ) : null}
-      {signUpState.stage === 'pick-handle' ? (
-        <PickHandlePage onNext={setSignUpState} />
-      ) : null}
-      {signUpState.stage === 'finish-profile' ? (
-        <FinishProfilePage onNext={setSignUpState} />
-      ) : null}
-      {signUpState.stage === 'select-genre' ? (
-        <SelectGenrePage onNext={setSignUpState} />
-      ) : null}
-      {signUpState.stage === 'select-artists' ? (
-        <SelectArtistsPage onNext={setSignUpState} />
-      ) : null}
-      {signUpState.stage === 'app-cta' ? <AppCtaPage /> : null}
+      <Switch>
+        <Route exact path={SIGN_UP_EMAIL_PAGE}>
```

Comment:
my expectations was that this would be the "SignupRoute" and each one would check if they are allowed to render, and if not, redirect? i think its probably fine to do this all in a root route like you did though.

Reply:
Yeah either way works technically.
I had moved it to the root since all the routes are using it; so I was thinking of it as a small possible optimization. 
However, I moved it down to each route and don't see any difference so I think any real optimization is probably super negligible 🤷 
Thinking about both I think I agree with you that wrapped around each route is probably a bit more readable since the logic is colocated so I'll keep it on the individual routes 👍

Reply:
Also I guess since we're probably rearranging it anyways, the individual wrappers makes it simpler to move around